[wallet] Add branch and bound coin selection

danielabrozzoni opened the pull request:
Fixes #48

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+
+// Calculated as:
+// prev_txid (32 bytes) + prev_vout (4 bytes) + script_len (1 bytes) + script len (0 bytes) + sequence (4 bytes)
+const TXIN_DEFAULT_WEIGHT: usize = 32 + 4 + 1 + 4;
```
From what I read this only supports native segwit inputs. Is that the intention?

danielabrozzoni replied:
Not exactly. This should be the weight needed to build the txin without the satisfaction weight, i.e., should be standard for every input. One thing this comment made me notice is that I shouldn't add `script_len` here, as that's part of the `satisfaction_weight` (and it's currently counted in every other CS implemented, I'll fix this in a separate commit) :)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+struct OutputGroup {
+    utxo: UTXO,
+    // Weight needed to spend the UTXO
+    weight: usize,
```
Judging from line 262, this is the weight of the witness only. If that's the case, I think this should be reflected in the comment.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+        &self,
+        _database: &D,
+        must_use_utxos: Vec<(UTXO, usize)>,
+        may_use_utxos: Vec<(UTXO, usize)>,
```
As discussed, these are being renamed to `required_utxos` and `optional_utxos`

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+        must_use_utxos: Vec<(UTXO, usize)>,
+        may_use_utxos: Vec<(UTXO, usize)>,
+        fee_rate: FeeRate,
+        amount_needed: u64,
```
`amount_needed` could also be more descriptive. I think this could be more accurately described as `sum_receiver_outputs`

danielabrozzoni replied:
I think this should be discussed in #146 :)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+        may_use_utxos: Vec<(UTXO, usize)>,
+        fee_rate: FeeRate,
+        amount_needed: u64,
+        fee_amount: f32,
```
Given that this seems to be the fees for the outputs and the transaction header, it would be more descriptive to call this `payload_fees` or `output_and_header-fees`. Otherwise, it might be nice to add a few comments here. If I didn't know the algorithm you're implementing, I think I'd be stumped to understand what's going on. :)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+            return Err(Error::InsufficientFunds);
+        }
+
+        if let Ok(c) = self.bnb(
```
This might be my novice level understanding of Rust speaking, but what does `Ok(c)` do here? I assume that it checks that the function finished successfully and might write the result into "c". If so, why is the "if-branch" also `Ok(c)` instead of just `c`? Shouldn't the result be given back rather than a check that the result was successful?

danielabrozzoni replied:
What's happening here is that I'm checking if the bnb was successful, and if it was, I'm returning `Ok(c)`, otherwise I'm returning `Ok(self.single_random_draw(...))`. The point is that `coin_select` returns a `Result`, so either `Ok(something)` or `Err(something)`.
Note that at this point of the code we are sure we are returning a successful result: we already checked that the funds are sufficient, so if the bnb fails we return `single_random_draw`, which doesn't even return a `Result<>`, just a `CoinSelectionResult` (which is a struct, and not a `Result<T, E>`. Do you think this could lead to any confusion?).
 
You could rewrite this as: (not very idiomatic, but I think it could help you understand :) )
```
let result = self.bnb(...);
if result.is_err() {
    result = Ok(self.single_random_draw(...)) // <- note that self.single_random_draw() doesn't return a Result, so we have to wrap it with Ok()
}
result
```
or, worse:
```
Ok(if let Ok(c) = self.bnb(...) { c } else { self.single_random_draw(...) }) // <- I'm basically wrapping either c or the single random draw in a Ok()
```

Your comment made me realize that this piece of code is super ugly and not very idiomatic though, so I'll refactor it as:

```
Ok(self.bnb(...).unwrap_or_else(|_| self.single_random_draw(...)))
```

Which basically means "if bnb is `Ok(c)` return c, otherwise return `single_random_draw(...)`", then wrap everything in a `Ok(...)`

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+        fee_amount: f32,
+        cost_of_change: f32,
+    ) -> Result<CoinSelectionResult, Error> {
+        let mut current_selection = Vec::with_capacity(may_use_utxos.len());
```
In case all required and all optional UTXO get used, wouldn't the length be ` may_use_utxos.len()+must_use_utxos.len()`?

murchandamus replied:
I realized later that the selection only works with the optional UTXO.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+            // Cannot possibly reach target with the amount remaining in the curr_available_value,
+            // or the selected value is out of range.
+            // Go back and try other branch
+            if curr_value + curr_available_value < actual_target
+                || curr_value > actual_target + cost_of_change as u64
+            {
+                backtrack = true;
```
I think this will cause a backtrack before any element is selected if the sum of the `must_use_utxos` already exceeds the target range. I haven't read your backtrack code yet, but that might throw. Please add that testcase if you don't have it yet.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+                    *c = false;
+                } else {
+                    // We have walked back to the first utxo and no branch is untraversed. All solutions searched
+                    break;
```
Nevermind, looks like the case with the required UTXO exceeding target should get caught here.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -238,6 +241,271 @@ impl<D: Database> CoinSelectionAlgorithm<D> for DumbCoinSelection {
+                while let Some(false) = current_selection.last() {
+                    current_selection.pop();
+                    curr_available_value +=
+                        may_use_utxos[current_selection.len()].effective_value as u64;
```
Is the `current_selection.len()` right here? If you are exploring a branch that has omitted multiple UTXO, I don't think the selection's length needs to match the position in the available UTXO anymore.

danielabrozzoni replied:
`current_selection` is a vector of booleans, and `current_selection[i]` contains `true` if we're using `may_use_utxos[i]`, `false` otherwise. (This is not very idiomatic Rust, I'm currently looking at a way to refactor it). 
So `current_selection.len()` is always `<=` than `may_use_utxos.len()`, but `current_selection[i]` matches `may_use_utxos[i]` `for i in 0..current_selection.len()`
For example:
```
current_selection: [true, false, false]
may_use_utxos (only the effective_values): [1, 2, 3, 4, 5]

if you current_selection.pop() you are removing current_selection[2],
and current_selection.len() becomes 2;
at this point you have to add may_use_utxos[2]
```

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -381,4 +682,152 @@ mod test {
+
+        assert_eq!(result.txin.len(), 2);
+        assert_eq!(result.selected_amount, 300_000);
+        assert_eq!(result.fee_amount, 186.0);
```
The BnB algorithm never allows change to be created. In the case that no changeless solution is found, it is supposed to fallback to another selection method.
Since the two available inputs are 100,000 and 200,000 satoshis, this should result in a failure for BnB. Since 50,000 sats is much larger than the fee for two inputs at 1 sat/vB and the permitted excess.

danielabrozzoni replied:
Yeah at the moment what's happening is that BnB is giving back TotalTriesExceeded (even if this is not exactly true: BnB tried everything and didn't find a suitable solution, it's not that there weren't enough tries!), single random draw is correctly giving back 300_000. 

I have updated the tests so they test the functions `bnb` and `single_random_draw` (instead of testing only `coin_select`)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -381,4 +682,152 @@ mod test {
+            )
+            .unwrap();
+
+        assert_eq!(result.txin.len(), 2);
+        assert_eq!(result.selected_amount, 300_000);
+        assert_eq!(result.fee_amount, 186.0);
```
See above.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -381,4 +682,152 @@ mod test {
+    // Every possible solution will be too big, and bnb will throw an error. Single random draw
+    // will be used, and only one utxo will be chosen.
+    #[test]
+    fn test_bnb_coin_selection_tries_exceeded() {
```
The name of this test seems to imply that you are testing the "tries exceeded", but since there are only 2 UTXOs, there are only 3 possible combinations.

To make a testcase for tries exceeded, you may want to create a UTXO Set that has a few sets of thousands of UTXOs with the same values. ;)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -381,4 +682,152 @@ mod test {
+                .unwrap();
+            assert_eq!(result.selected_amount, target_amount);
+        }
+    }
```
This is a good sanity check. I would like to propose another. Generate 200 random UTXOs and make the target the sum of the effective value of two of them. You should only get a results with one or two imputs from that. I think that might uncover at least one bug in the above code (I think there is an issue in the backtrack). :)

Alternatively, I would like to suggest a test with a known set of available UTXOs where it e.g. uses the first and fifth largest UTXO of the set to construct the target.

danielabrozzoni replied:
I wrote the first test you suggested and it's unsurprisingly failing (TotalTriesExceeded), but I don't get why it should pass. 200 UTXOs are really a lot (with values ranging from 0 sats to 2BTC), how could the bnb find the exact match with only 100_000 tries?

danielabrozzoni replied:
What is the complexity of the BnB approach? `O(n^2)`? (If so, it should definitely pass :))

murchandamus replied:
The whole search space of BnB is O(2^n), but the bounding essentially makes it go to linear complexity for some parts. Finding an exact match with two inputs can be done in less than n^2 tries though. I think we will need a few more bounding mechanisms. :)

murchandamus reviewed (changes requested):
This is coming along nicely! I think the tests with two UTXO are a bit too easy to actually uncover some of the issues around Branch and Bound. I would like to suggest that you increase the available UTXO at least into the range of 10 to see some interesting behaviors. Also, since the target window is feerate dependent, you might want to add a couple more tests with hier feerates.

Personally, I would prefer if a PR this large was composed of multiple commits. A few spots could use a bit more explanation, I think it may be hard to follow along in a few spots if one isn't as familiar with the intended outcome as myself.

danielabrozzoni committed:
[wallet] Add Branch and Bound coin selection

danielabrozzoni committed:
[wallet] Add tests for BranchAndBoundCoinSelection::coin_select

danielabrozzoni committed:
[wallet] Add tests for BranchAndBoundCoinSelection::bnb

danielabrozzoni committed:
[wallet] Add tests for BranchAndBoundCoinSelection::single_random_draw

danielabrozzoni committed:
[wallet] Replace `must_use` with `required` in coin selection

danielabrozzoni commented:
I've updated the code, should be little bit better now. As I was saying in the comment, the test with 200 UTXOs is currently failing (throws TotalTriesExceeded). I'll dig into it :)

danielabrozzoni commented:
I updated the PR so it's not a draft anymore.
I would say: the fact that bnb can't handle more than 40/45 utxos is an issue... but maybe we shouldn't focus too much on it now. I added a `TODO:` so we don't forget that in the future some optimizations will be needed :)

murchandamus commented:
Okay, I'll review it again. :)

danielabrozzoni committed:
[wallet] Use TXIN_DEFAULT_WEIGHT constant in coin selection

Replace all the occurences of `serialize(&txin)`
with TXIN_DEFAULT_WEIGHT.

afilini reviewed (approved):
Even though there are a few things that could probably be expressed in a more "rust-like" style, I would go ahead and merge this so that we can prepare a tag for the next beta release, and then work on more improvements in separate PRs.

Do you have any more comments @Xekyo before I go ahead and merge this?

LLFourn commented:
I've made a smaller PR [here] that will conflict a bit with this. All in all it looks easier to merge that one first and then this one. I don't mind updating this PR if that works. Not a big issue either way.

As a side note I think it would be good if the B&B coin selection algorithm go in its own file in a coin_selection directory. Presumably there could be more than one coin selection algorithm going forward.

[here]: https://github.com/bitcoindevkit/bdk/pull/159

danielabrozzoni commented:
I'm cool with merging #159 before this one :)
I like the idea of having a coin_selection directory with different files for the different algos, but I would prefer to do this refactoring in a different PR, so it's easier to review

LLFourn commented:
@danielabrozzoni, @afilini  it looks like there are some complications to that PR that requires some further thinking so I'd say go for this one :)

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -240,6 +242,292 @@ impl<D: Database> CoinSelectionAlgorithm<D> for LargestFirstCoinSelection {
 
+#[derive(Debug, Clone)]
+// Adds fee information to an UTXO.
+struct OutputGroup {
+    utxo: UTXO,
+    // weight needed to satisfy the UTXO, as described in `Descriptor::max_satisfaction_weight`
+    satisfaction_weight: usize,
+    // Amount of fees for spending a certain utxo, calculated using a certain FeeRate
+    fee: f32,
+    // The effective value of the UTXO, i.e., the utxo value minus the fee for spending it
+    effective_value: i64,
+}
```
Just a bit background info: `OutputGroup`s in Bitcoin Core collect all unspents associated with one address. They are used to prevent spending only one UTXO if there are multiple associated with the same address, and to rather only spend all or none of them, so that there is no additional privacy leak.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -240,6 +242,292 @@ impl<D: Database> CoinSelectionAlgorithm<D> for LargestFirstCoinSelection {
+            .collect();
+
+        // Mapping every (UTXO, usize) to an output group.
+        // Filtering UTXOs with an effective_value < 0, as the fee paid for
```
Nit: Strictly speaking, I think you're filtering `effective_value <= 0`.

murchandamus commented on src/wallet/coin_selection.rs:
```diff
@@ -856,4 +856,135 @@ mod test {
+        let mut rng: StdRng = SeedableRng::from_seed(seed);
+        let fee_rate = FeeRate::from_sat_per_vb(0.0);
+
+        for _ in 0..200 {
```
Am I understanding correctly, that you're performing this test 200 times? Would that perhaps be the reason why it timed out with a larger number of UTXOs? If I understand that correctly, have you tried running it with ~500 UTXOs just once?

murchandamus commented:
One optimization that I had added was to skip a UTXO's inclusion branch when it exactly matches the one that was just backtracked. E.g.

A = 100
B = 90
C = 90
D = 80
…

Would then have the exploration: { [A], [AB], [ABC], [ABCD], …, [ABC_…], [AB__…], A__D…]. I.e. once B gets omitted, C gets skipped, because the prefix [AB…] and [AC…] would be equivalent.

The observation was that people often send round values and this was an easy way to skip a bunch of equivalent solutions.
